packages/engine/crawlers/solana/transactionHistory.ts

```diff
@@ -0,0 +1,530 @@
+		if (!tokenBalance.owner) continue;
+
+		if (
+			tokenBalance.owner !== FEE_PAYER &&
+			tokenBalance.owner !== (type === 'sent' ? sender : receiver)
+		) {
+			const network = Networks.solana;
+			const account: TokenAccount = {
+				mint: tokenBalance.mint,
+				owner: tokenBalance.owner,
+				address: (instruction as ParsedInstruction).parsed?.info.source,
+				balance: tokenBalance.uiTokenAmount.uiAmount?.toString() || '',
+				decimals: tokenBalance.uiTokenAmount.decimals,
+			};
+			const metadata = await getMetadata(solanaContext, tokenBalance.mint);
+
+			token = {
+				network,
+				account,
+				metadata,
+			};
+
+			return token;
+		}
+	}
+
+	return token;
+};
+
+const getTransactionBalances = (
+	meta: ParsedTransactionWithMeta['meta'],
+	ownerPublicKey: string,
+	type?: 'sent' | 'received',
+) => {
+	if (!meta) {
+		return {
+			preBalance: 0,
+			postBalance: 0,
+		};
+	}
+	if (
+		!meta.preTokenBalances ||
+		!meta.postTokenBalances ||
+		meta.postTokenBalances.length === 0 ||
+		meta.preTokenBalances.length === 0
+	) {
+		if (type === 'sent') {
+			const preBalance = meta.preBalances[0] / 10 ** 9;
+			const postBalance = meta.postBalances[0] / 10 ** 9;
+
+			return {
+				preBalance,
+				postBalance,
+			};
+		}
+
+		return {
+			preBalance: meta.preBalances[1] / 10 ** 9,
+			postBalance: meta.postBalances[1] / 10 ** 9,
+		};
+	} else {
+		let preBalance = 0;
+		let postBalance = 0;
+
+		const preBalances = meta.preTokenBalances.filter(
+			(item) => item.owner === ownerPublicKey,
+		);
+		const postBalances = meta.postTokenBalances.filter(
+			(item) => item.owner === ownerPublicKey,
+		);
+
+		if (!preBalances[0]?.uiTokenAmount.uiAmount) {
+			preBalance = 0;
+		} else {
+			preBalance = preBalances[0].uiTokenAmount.uiAmount;
+		}
+		if (!postBalances[0]?.uiTokenAmount.uiAmount) {
+			postBalance = 0;
+		} else {
+			postBalance = postBalances[0].uiTokenAmount.uiAmount;
+		}
+
+		return {
+			preBalance,
+			postBalance,
+		};
+	}
+};
+
+const getTransactionAmount = (
+	token: Token,
+	tokeForFee: Token,
+	instruction: ParsedInstruction | PartiallyDecodedInstruction,
+) => {
+	const parsedInstructionInfo = (instruction as ParsedInstruction).parsed?.info;
+
+	let amount = 0;
+
+	if (parsedInstructionInfo?.lamports) {
+		amount = parsedInstructionInfo?.lamports / 10 ** 9;
+	}
+
+	if (parsedInstructionInfo?.amount) {
+		amount = parsedInstructionInfo?.amount / 10 ** token.account.decimals;
+	}
+
+	if (parsedInstructionInfo?.tokenAmount?.amount) {
+		amount =
+			parseFloat(parsedInstructionInfo?.tokenAmount.amount) /
+			10 ** token.account.decimals;
+	}
+
+	if (token.account.mint === tokeForFee.account.mint) {
+		amount = parsedInstructionInfo?.amount / 10 ** token.account.decimals;
+	}
+
+	if (token.account.mint === solMint) {
+		amount = parsedInstructionInfo?.lamports / 10 ** 9;
+	}
+
+	if (token.metadata?.mpl) {
+		amount = 1;
+	}
+
+	return amount;
+};
+
+const getTransactionType = (
```

Comment:
a transaction might have both received and sent instructions. I think we need to redefine this one. Also I see a lot of duplicate condition checking, consider to merge these code with same condition checking